provider-example/src/aead.rs

```diff
@@ -187,21 +175,45 @@ impl cipher::MessageDecrypter for Tls12Cipher {
 
 const CHACHAPOLY1305_OVERHEAD: usize = 16;
 
-struct BufferAdapter<'a, 'p>(&'a mut BorrowedPayload<'p>);
+struct EncryptBufferAdapter<'a>(&'a mut PrefixedPayload);
+
+impl AsRef<[u8]> for EncryptBufferAdapter<'_> {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl AsMut<[u8]> for EncryptBufferAdapter<'_> {
+    fn as_mut(&mut self) -> &mut [u8] {
+        self.0.as_mut()
+    }
+}
+
+impl Buffer for EncryptBufferAdapter<'_> {
+    fn extend_from_slice(&mut self, other: &[u8]) -> chacha20poly1305::aead::Result<()> {
+        (&mut self.0 as &mut Vec<u8> as &mut dyn Buffer).extend_from_slice(other)
+    }
+
+    fn truncate(&mut self, len: usize) {
+        (&mut self.0 as &mut Vec<u8> as &mut dyn Buffer).truncate(len)
+    }
+}
+
+struct DecryptBufferAdapter<'a, 'p>(&'a mut BorrowedPayload<'p>);
 
-impl AsRef<[u8]> for BufferAdapter<'_, '_> {
+impl AsRef<[u8]> for DecryptBufferAdapter<'_, '_> {
```

Comment:
Similarly, I think it would be better to leave `BufferAdapter` with the same name in the commit that adds `EncryptBufferAdapter`, and rename it in the commit after that.

Reply:
Done.